server/src/main/java/com/thoughtworks/go/server/service/EntityHashes.java

```diff
@@ -0,0 +1,144 @@
+import static com.thoughtworks.go.util.CachedDigestUtils.sha512_256Hex;
+import static org.apache.commons.lang3.StringUtils.join;
+
+@Component
+public class EntityHashes {
+    public static final String SEP_CHAR = "/";
+
+    private static final Gson GSON = new GsonBuilder().
+            registerTypeAdapter(ConfigurationProperty.class, Serializers.CONFIGURATION_PROPERTY).
+            registerTypeAdapter(EnvironmentVariableConfig.class, Serializers.ENVIRONMENT_VARIABLE).
+            registerTypeAdapter(PluginInfo.class, Serializers.PLUGIN_INFO).
+            create();
+
+    private final MagicalGoConfigXmlWriter xmlSerializer;
+
+    @Autowired
+    public EntityHashes(ConfigCache configCache, ConfigElementImplementationRegistry registry) {
+        xmlSerializer = new MagicalGoConfigXmlWriter(configCache, registry);
+    }
+
+    private static String digestHex(String data) {
+        return sha512_256Hex(data);
+    }
+
+    /**
+     * Computes a digest of a {@link PartialConfig}.
+     *
+     * @param partial a {@link PartialConfig} to hash
+     * @return a cryptographic digest that can be used for comparison.
+     */
+    public String digestPartial(PartialConfig partial) {
+        if (null == partial) {
+            return null;
+        }
+
+        return digestMany(
+                digestMany(partial.getGroups()),
+                digestMany(partial.getEnvironments()),
+                digestMany(partial.getScms())
+        );
+    }
+
+    public String digestMany(String... data) {
+        return digestHex(join(data, SEP_CHAR));
+    }
+
+    /**
+     * Computes a cryptographic digest of a collection's contents
+     *
+     * @param entities any {@link Collection} of config entities
+     * @return a cryptographic hex digest ({@link String})
+     */
+    public String digestMany(Collection<?> entities) {
+        if (null == entities) {
+            return null;
+        }
+
+        return digestHex(entities.stream().
+                map(this::digestDomainConfigEntity).
+                collect(Collectors.joining(SEP_CHAR)));
+    }
+
+    public String digestDomainConfigEntity(Object entity) {
+        return digestHex(serializeDomainEntity(entity));
+    }
+
+    public String digestDomainNonConfigEntity(Object entity) {
+        return digestHex(GSON.toJson(entity));
+    }
+
+    protected String serializeDomainEntity(Object domainObject) {
+        return xmlSerializer.toXmlPartial(domainObject);
+    }
+
+    private interface Serializers {
+        /**
+         * Custom serializer for encrypted data to ensure stable JSON output when crypto salt changes
+         * over certain requests due to re-encryption
+         */
+        JsonSerializer<ConfigurationProperty> CONFIGURATION_PROPERTY = (src, typeOfSrc, context) -> {
+            final JsonObject result = new JsonObject();
+            result.addProperty("key", src.getConfigKeyName());
+            result.addProperty("value", sha512_256Hex(src.getValue()));
```

Comment:
@maheshp I think I did this mostly out of paranoia; in case anyone ever exposes the `Gson` instance without thinking about consequences, I didn't want actual secrets ever making it out to I/O (`getValue()` may return decrypted secrets if `src` is an encrypted value, IIRC). It's not _really_ necessary. If you think performance will be negatively impacted (I don't, but you may know that something I don't), I can change it to be plaintext.